### ISSUE TITLE
How can I improve my workflow to obtain accurate inference results using Pytorch Hub ?

### ISSUE THREAD

**Issue by bzimon** [ultralytics/yolov5#8575]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello,

I am working with quite big microscopy images (100k x 50k) roughly. Because of that I trained yolo using 640 x 640 chunks - around 1500 images were used for training. The model performed well on unseen images from different experiments (more than 95% of precision and recall). 

Due to large size of the slide I switched from detect.py to Pytorch Hub, and applied the model to numpy arrays extracted in a sliding window fashion. 

My problem is that bounding boxes do not seem to make sense. All our labels used in the training were exactly the same size (50x50) because objects we detect are constant size as are the images. Pytorch hub inference resulted in most of the boxes of very different size, and in wrong locations. My input images for training where single channel tiff files (array of uint16). I used iou = 0.5 and confidence threshold = 0.3 those values shown the best results during testing.

uint16 was not compatible with the Hub so i converted to int16, I also tried int32, instead. Do you think this is a root cause of the problem ? 
Do you have any suggestions on how to improve the detection?

I attach very good detections found during testing and a bad ones generated using hub. 
**High quality detection using detect.py**
![image](https://user-images.githubusercontent.com/39350665/178805025-9daa8116-1086-4879-9773-6a279610ba64.png)

**Incorrect detections using Pytorch Hub**
![image](https://user-images.githubusercontent.com/39350665/178805496-943a4081-2b30-43d9-9697-dd69f8b16857.png)
![image](https://user-images.githubusercontent.com/39350665/178807443-767f8257-c8ed-46f6-bbf1-11751949392b.png)

I would be grateful for any suggestions, I am sure i missed something, but cannot figure out what.
Thank you for development of yolov5 and looking forward to your suggestions! 

**Custom trained model loading**
````
model = torch.hub.load(r"\yolov5",
                       'custom',
                       path=r"best.pt",
                       source='local').eval()
model.conf = 0.3  # NMS confidence threshold
model.iou = 0.5  # NMS IoU threshold

for param in model.parameters():
    param.grad = None
````

**Inference with the Hub**

``` 
all_detections = pd.DataFrame(columns = ['g_xmin', 'g_xmax', 'g_ymin', 'g_ymax', 'xmin', 'ymin', 'xmax', 'ymax', 'confidence', 'class', 'name'])

img_entry = AICSImage(img_iter[0].uri.replace('file:/',''))
img_arr = img_entry.xarray_dask_data[0,0,0,:,:] # TCZYX
mem_img = img_arr.data.compute().astype(np.int32())

for grids in tqdm(grids_lst):
    g_xmin,g_xmax,g_ymin,g_ymax = grids
    imSlice = mem_img[g_ymin : g_ymax,g_xmin : g_xmax]  #getting slices from microscopy image
    with torch.no_grad():
        res = model(imSlice).pandas().xyxy[0]
        if not res.empty:
            
            #Combining detections and grid coordinates
            res_df = pd.merge(res,
                           pd.DataFrame([[g_xmin,  g_xmax,  g_ymin, g_ymax, g_xmax]], columns = ['g_xmin',  'g_xmax',  'g_ymin', 'g_ymax', 'g_xmax']),
                           how='cross')
            all_detections.reset_index(drop = True, inplace = True)
            res_df.reset_index(drop=True, inplace=True)
            if len(all_detections) > 0: 
                all_detections = pd.concat((all_detections,res_df), axis=0)
            else:
                all_detections = res_df
                
            #Ploting detections
            fig, ax = plt.subplots(figsize=(8, 6))
            for dtc in res_df.iterrows():
                indx, dtcDf = dtc
                rect = patches.Rectangle(xy = (int(dtcDf['xmin']), int(dtcDf['ymin'])),
                                         width = int(dtcDf['xmax'] - dtcDf['xmin']),
                                         height = int(dtcDf['ymax'] - dtcDf['ymin']), 
                                         linewidth=1,
                                         edgecolor='cyan',
                                         facecolor='none')
                ax.add_patch(rect)
            ax.set_title(str(g_xmin) + '_' + str(g_xmax))
            ax.imshow(imSlice, 'Reds_r')
            plt.show()   
```
**All detections dataframe sample**
![image](https://user-images.githubusercontent.com/39350665/178808164-8d463dec-efca-4873-a415-921c0750476c.png)

**Comment by bzimon** [comment]
My problem was improper array rescaling it works now. Hope this question will help someone else in the same situation.
```
 interp = np.interp(imSlice, (imSlice.min(), imSlice.max()), (0, 7)).astype(np.uint8)
 rshapedImg, ratio, (dw, dh) = letterbox(interp)
 res = model(rshapedImg , size=640).pandas().xyxy[0]
```

**Comment by glenn-jocher** [comment]
@bzimon yes you should get identical results from detect.py and PyTorch Hub, if not likely your preprocessing is incorrect for the PyTorch Hub inputs.

**Comment by lioneldaniel** [comment]
@bzimon Did you succeed in training your model on the full image dynamic, from 0 to 65535, instead of from 0 to 255 when images are converted to `uint8` by opencv ?
The source code is using [`cv2.imread(f)`](https://github.com/ultralytics/yolov5/blob/master/utils/dataloaders.py#L700), which returns a numpy.ndarray of dtype `uint8`, whereas `cv2.imread(f, cv2.IMREAD_UNCHANGED)` would output a numpy.ndarray of dtype `uint16`

**Comment by cirqleBio** [comment]
@lioneldaniel I didn't try again with the full range. The model did ok after rescaling. Though I think there could be some advantage with uint16.

**Comment by bzimon** [comment]
Hey, I did not try with the full range, because the model performed well enough after rescalling. Though if you succeed with uint16 let me know, I can see a potential advantage there.
